controllers/resources/lease.go

```diff
@@ -0,0 +1,249 @@
+	// InvalidateLease extends or releases a lease based on the CR status, type of remediation and how long the lease is already leased
+	InvalidateLease(ctx context.Context, remediationCR *unstructured.Unstructured) error
+}
+
+type nhcLeaseManager struct {
+	client             client.Client
+	commonLeaseManager lease.Manager
+	log                logr.Logger
+}
+
+func NewLeaseManager(client client.Client, log logr.Logger) (LeaseManager, error) {
+	newManager, err := lease.NewManager(client, holderIdentity)
+	if err != nil {
+		log.Error(err, "couldn't initialize lease manager")
+		return nil, err
+	}
+	return &nhcLeaseManager{
+		client:             client,
+		commonLeaseManager: newManager,
+		log:                log.WithName("nhc lease manager"),
+	}, nil
+}
+
+func (m *nhcLeaseManager) ObtainNodeLease(remediationCR *unstructured.Unstructured, nhc *remediationv1alpha1.NodeHealthCheck) (bool, *time.Duration, error) {
+	nodeName := remediationCR.GetName()
+	leaseDuration := m.getLeaseDurationForRemediation(remediationCR, nhc)
+	leaseDurationWithBuffer := leaseDuration + LeaseBuffer
+
+	node := &v1.Node{}
+	if err := m.client.Get(context.Background(), types.NamespacedName{Name: nodeName}, node); err != nil {
+		m.log.Error(err, "couldn't obtain node lease node error getting node", "node name", nodeName)
+		return false, nil, err
+	}
+
+	if err := m.commonLeaseManager.RequestLease(context.Background(), node, leaseDurationWithBuffer); err != nil {
+		if _, ok := err.(lease.AlreadyHeldError); ok {
+			m.log.Info("can't acquire node lease, it is already owned by another owner", "already held error", err)
+			return false, &RequeueIfLeaseTaken, err
+		}
+
+		m.log.Error(err, "couldn't obtain lease for node", "node name", nodeName)
+		return false, nil, err
+	}
+
+	//all good lease created with wanted duration
+	return true, &leaseDuration, nil
+
+}
+
+func (m *nhcLeaseManager) ManageLease(ctx context.Context, remediationCR *unstructured.Unstructured, nhc *remediationv1alpha1.NodeHealthCheck) (time.Duration, error) {
+	node := &v1.Node{}
+	if err := m.client.Get(ctx, client.ObjectKey{Name: remediationCR.GetName()}, node); err != nil {
+		m.log.Error(err, "managing lease - couldn't fetch node", "node name", remediationCR.GetName())
+		return 0, err
+	}
+	l, err := m.commonLeaseManager.GetLease(ctx, node)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return 0, nil
+		}
+		m.log.Error(err, "managing lease - couldn't fetch lease", "node name", remediationCR.GetName())
+		return 0, err
+	}
+
+	if exist := remediationCR.GetDeletionTimestamp() == nil; !exist && m.isLeaseOwner(l) {
+		m.log.Info("managing lease - lease has no remediations so  about to be removed", "lease name", l.Name)
+		//release the lease - no remediations
+		return 0, m.commonLeaseManager.InvalidateLease(ctx, node)
+	} else if ok, err := m.isLeaseOverdue(l, nhc, remediationCR); err != nil {
```

Comment:
My turn for a nit :-)

The `else`'s are redundant when every block returns

Reply:
I agree about the last `else`- it's not that small and we can cut off one nesting level by doing that

regarding the `else if's` I prefer to keep as is:
- I don't think they add little to no complexity in the code (over the alternative of just `if`)
- using that patter allow defining the variables withing the cond block and using it across all cases - IMO it is much nicer than having them predefined  (especially for such a use case with very small blocks)